MC-17694: Inventory Threshold doesn't display for configurable produc…

YuliyaLabudova opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
* [MC-17694](https://jira.corp.magento.com/browse/MC-17694) Inventory Threshold doesn't display for configurable products simple product

### Related Pull Requests
https://github.com/magento-epam/magento2ce/pull/221 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Add threshold qty for msi.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Add threshold qty for msi.

ishakhsuvarov commented:
Hi @YuliyaLabudova 

Could you please describe how this feature is supposed to work? I've tried creating a configurable product composed of simple products with qty of 1 with `Only X left` setting set to 100. I was unable to see any changes to the product page. 

It would be best if the new feature was accompanied by description, test cases and automated tests.

YuliyaLabudova commented:
> Hi @YuliyaLabudova
> 
> Could you please describe how this feature is supposed to work? I've tried creating a configurable product composed of simple products with qty of 1 with `Only X left` setting set to 100. I was unable to see any changes to the product page.
> 
> It would be best if the new feature was accompanied by description, test cases and automated tests.

Hi, we have already done it and add the mftf test for it(please, see the ticket: https://jira.corp.magento.com/browse/MC-17694, I've attached the green build for the ticket).
But we have changes in ce also: https://fisheye.corp.magento.com/cru/CR-MAGETWO-46918
It works along with these changes only. 
I am not sure how we should deliver it now, but last time we delivered changes to msi and ce separately
Could you, please, advice how should we deliver it?

ishakhsuvarov commented:
Tests and CE branch were not linked before, PR looked like it's only targeting MSI. 

What is the purpose of Inventory Pull Request if it's also delivered to CE? Does this specific Pull Request provide multi stock support for the feature being delivered? Are there any test scenarios?

YuliyaLabudova commented:
> Tests and CE branch were not linked before, PR looked like it's only targeting MSI.
> 
> What is the purpose of Inventory Pull Request if it's also delivered to CE? Does this specific Pull Request provide multi stock support for the feature being delivered? Are there any test scenarios?

Changes for CE still is in review status but we create PR only for reviewed changes.
The feature should have been delivered to CE, but the mftf test was failed with MSI extension in jenkins.
After discussion with @kieuphan2406, it was decided to fix it for MSI. 
This PR provides multi stock support for the feature being delivered.
About test scenarios: we have the test in CE for it. The main difference: product in MSI created with quantity for default stock already. Thus, the test covers the scenario described in the test case(https://jira.corp.magento.com/browse/MC-17951) with default stock. But we can extend the test for MSI if it needed. Should we do that?

Actually, I cannot answer the question about linked PR for CE, because the test for the ticket will break down because of existing teams use MSI release branch and, I think, It's better to create the ticket for skipping the test after merging these changes to MSI.

ishakhsuvarov commented:
@YuliyaLabudova Ok, we can deliver these in the same time, if that helps. First we would have to:
* Complete review of the CE commits
* Adjust mftf tests for MSI if needed. After reading comment above it's still unclear to me.
* Complete review & testing with MSI.
* Add appropriate description of the feature and it's effect in MSI to this ticket. We have to cover it with Publicly available test case and have reliable test scenario before delivery.

I'll be tracking MC-17694 to make sure we are aligned.

maghamed commented on app/code/Magento/InventoryConfigurableProduct/Plugin/Block/Product/View/Type/AddOptionQtysPlugin.php:
```diff
@@ -0,0 +1,111 @@
+    private function getOptionQtys(Subject $configurable): array
+    {
+        $qtys = [];
+        foreach ($configurable->getAllowProducts() as $product) {
```
this operation may potentially lead to performance degradation for Configurable products with many children assigned. 
As for each and any child, we proceed with 2 independent service calls.  Configuration and Salable Quantity calculation

maghamed commented on app/code/Magento/InventoryCatalog/Model/GetProductQtyById.php:
```diff
@@ -0,0 +1,83 @@
+        try {
+            $qty = $this->getProductSalableQty->execute($sku, $stockId);
+        } catch (InputException $e) {
+            $qty = 0.0;
```
incorrect handling of exceptions

YuliyaLabudova replied:
Thank you, fixed

ishakhsuvarov replied:
What was the fix?

maghamed commented on app/code/Magento/InventoryConfigurableProduct/Plugin/Block/Product/View/Type/AddOptionQtysPlugin.php:
```diff
@@ -0,0 +1,111 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryConfigurableProduct\Plugin\Block\Product\View\Type;
```
the module where this logic should belong to should be `InventoryConfigurableProductFrontendUi`

maghamed commented on app/code/Magento/InventoryConfigurableProduct/Plugin/Block/Product/View/Type/AddOptionQtysPlugin.php:
```diff
@@ -0,0 +1,111 @@
+            || $this->stockItemConfig->getBackorders() !== StockItemConfigurationInterface::BACKORDERS_NO
+            && $this->stockItemConfig->getMinQty() < 0)
+            && $productSalableQty <= $this->stockItemConfig->getStockThresholdQty()
+            && $productSalableQty > 0;
```
this looks like a duplication of logic we have in \Magento\InventorySalesFrontendUi\Plugin\Block\Stockqty\AbstractStockqtyPlugin

maghamed reviewed (changes requested):
In general, this functionality better to incorporate via async JS call after the main page is loaded.
Because potentially for Configurable products with many children it may introduce performance degradation, as any child is handled independently and for each child, we need to handle 2 service calls along with additional business logic

YuliyaLabudova committed:
Merge branch '2.3-develop' of https://github.com/magento/inventory into MC-17694

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Add qty left for configurable variation by ajax request.

YuliyaLabudova committed:
Merge branch 'MC-17694' of https://git.epam.com/magn-m23/msi into MC-17694

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Fix static tests.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Move logic to Catalog Inventory.

YuliyaLabudova reviewed (commented):
> In general, this functionality better to incorporate via async JS call after the main page is loaded.
> Because potentially for Configurable products with many children it may introduce performance degradation, as any child is handled independently and for each child, we need to handle 2 service calls along with additional business logic

Thank you for your comments.
I changed the implementation for the ticket.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Add logic with threshold quantity for configurable quantity.

YuliyaLabudova committed:
Merge branch '2.3-develop' of https://github.com/magento/inventory into MC-17694

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Update content-hash for composer.lock.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Fix static tests.

irenelagno commented on app/code/Magento/InventoryCatalog/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,86 @@
+        $resultJson = $this->resultPageFactory->create(ResultFactory::TYPE_JSON);
+        $qty = $this->productQty->getProductQtyLeft($productId, (int)$this->storeManager->getStore()->getWebsiteId());
+
+        if ($qty === null) {
```
can we define qty and then do $resultJson->setData($qty)?

irenelagno commented on app/code/Magento/InventoryCatalog/Model/GetProductQtyById.php:
```diff
@@ -0,0 +1,83 @@
+     * @param int $scopeId
+     * @return float
+     */
+    public function execute(
```
please add type hinting for input parameters

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,86 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Controller\Product;
```
Please consider moving this controller to the `InventoryCatalogFrontendUi` module

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /**
+     * Product view action
```
Please keep comments as meaningful as possible. Would be great to describe exact purpose of this controller

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Get product qty left.
+ */
+class GetQty extends Action implements HttpGetActionInterface
```
To discuss: It might make sense to use Web API instead of controller for this purpose.

YuliyaLabudova replied:
As I know, we cannot add new webapi routes due to backward compatibility. 
Because now we haven't the same functionality in webapi (with checker), it was the reason for adding a new controller. Should we leave it as it is?

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Model/GetProductQtyById.php:
```diff
@@ -0,0 +1,83 @@
+     * @return float
+     */
+    public function execute(
+        $productId,
```
It might make sense to accept sku instead of product ID in this service ID -> SKU conversion may be done in controller.

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Model/GetProductQtyById.php:
```diff
@@ -0,0 +1,83 @@
+        $scopeId = null
+    ): float {
+        $websiteCode = $this->storeManager->getWebsite($scopeId)->getCode();
+        $stockId = $this->stockResolver->execute(SalesChannelInterface::TYPE_WEBSITE, $websiteCode)->getStockId();
```
We should not hardcode Sales Channel type here. It should be accepted as a parameter to the service

Service should accept `SalesChannelType` and `SalesChannelCode` as arguments.

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Model/GetProductQtyById.php:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * Get product qty by sku.
```
Please update comment to reflect real purpose.

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Model/ProductQty.php:
```diff
@@ -0,0 +1,77 @@
+     */
+    private function getStockQtyLeft(float $productSalableQty): float
+    {
+        return (float)($productSalableQty - $this->stockItemConfig->getMinQty());
```
This is already accounted by `\Magento\InventorySalesApi\Api\GetProductSalableQtyInterface` from which the Qty was obtained.

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Model/ProductQty.php:
```diff
@@ -0,0 +1,77 @@
+     * @param int $websiteId
+     * @return float|null
+     */
+    public function getProductQtyLeft(int $productId, int $websiteId):? float
```
We tend to use `execute` method name for such kind of services.

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/README.md:
```diff
@@ -0,0 +1,16 @@
+# InventoryConfigurableProductFrontend module
```
Please describe real purpose of the module. Currently it seems to be same as in other modules, without any specifics on the functionality.

YuliyaLabudova replied:
Fixed

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/composer.json:
```diff
@@ -0,0 +1,21 @@
+    "name": "magento/module-inventory-configurable-product-frontend",
+    "description": "N/A",
+    "require": {
+        "php": "~7.1.3||~7.2.0||~7.3.0",
```
Some dependencies seem to be missing

YuliyaLabudova replied:
Yes, I agree. But after adding dependencies static tests failed because we didn't use this functionality. Can I add this dependencies and ignore static tests?

irenelagno replied:
@YuliyaLabudova Could you fix this error?

YuliyaLabudova replied:
This problem is fixed only if I remove these dependencies

ishakhsuvarov commented on app/code/Magento/InventorySalesFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -37,25 +41,36 @@ class AbstractStockqtyPlugin
         $this->stockByWebsiteId = $stockByWebsiteId;
         $this->getProductSalableQty = $getProductSalableQty;
         $this->isSourceItemManagementAllowedForProductType = $isSourceItemManagementAllowedForProductType;
+        $this->qtyLeftChecker = $qtyLeftChecker ?: ObjectManager::getInstance()
```
Plugins are not covered by Backwards Compatibility policy. Please inject dependencies in a normal way.

ishakhsuvarov commented on app/code/Magento/InventorySwatches/etc/module.xml:
```diff
@@ -0,0 +1,10 @@
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
+    <module name="Magento_InventorySwatches" />
```
I'd suggest naming the module in the lines of `SwatchesFrontendUi`

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Model/ProductQty.php:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Model for getting product qty left.
+ */
+class ProductQty
```
Please clarify service name. It's not clear what is it's purpose.

ishakhsuvarov reviewed (changes requested):
Please consider the following:
* Possibly it's better to use Web API instead of controller.
* Please review service names an comments. Make sure those are clear.
* Please do not rely on Website directly. Use Sales Channel entity instead.

irenelagno commented on app/code/Magento/InventoryConfigurableProductFrontend/view/frontend/web/js/configurable-variation-qty.js:
```diff
@@ -0,0 +1,49 @@
+    'use strict';
+
+    return function (productId) {
+        const selectorInfoStockSkuQty = '.availability.only',
```
Replace `const`, `let` with `var` because we are still supporting IE 11

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Add new module;
- Fix js (remove let and const);
- Fix comments.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Fix static tests.

irenelagno commented on app/code/Magento/InventoryCatalogFrontendUi/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        $sku = $this->getSkusByProductIds->execute([$productId])[$productId];
+        $websiteCode = $this->storeManager->getWebsite((int)$this->storeManager->getStore()->getWebsiteId())->getCode();
```
please simplify this code. See implementation of storeManager->getWebsite

irenelagno commented on app/code/Magento/InventoryCatalogFrontendUi/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,103 @@
+            return $resultForward;
+        }
+
+        $sku = $this->getSkusByProductIds->execute([$productId])[$productId];
```
is exception case handled properly on frontend?

irenelagno commented on app/code/Magento/InventoryConfigurableProductFrontend/README.md:
```diff
@@ -0,0 +1,16 @@
+# InventoryConfigurableProductFrontend module
+
+The `InventoryConfigurableProductFrontend` add info about qty left quantity for configurable variations.
```
please fix grammatic error: add->adds

irenelagno commented on app/code/Magento/InventorySwatchesFrontendUi/README.md:
```diff
@@ -0,0 +1,13 @@
+# InventorySwatchesFrontendUi module
+
+The `InventorySwatchesFrontendUi` add info about qty left to ui pages.
```
adds

irenelagno commented on app/code/Magento/InventorySwatchesFrontendUi/Test/Mftf/Test/StorefrontDisplayCorrectMessageWhenChildProductSelectedForConfigProductTest.xml:
```diff
@@ -0,0 +1,304 @@
+        </annotations>
+        <before>
+            <actionGroup ref="LoginAsAdmin" stepKey="loginAsAdmin"/>
+            <magentoCLI command="config:set  cataloginventory/options/stock_threshold_qty 2" stepKey="setQty2"/>
```
xample of good mftf fixture https://github.com/magento-epam/magento2ce/blob/2.3-develop/app/code/Magento/Captcha/Test/Mftf/Data/CaptchaConfigData.xml 
than you can use in the test such format

`<magentoCLI command="config:set {{StorefrontEnableAddStoreCodeToUrls.path}} {{StorefrontEnableAddStoreCodeToUrls.value}}" stepKey="addStoreCodeToUrlEnable"/>`

irenelagno commented on app/code/Magento/InventorySwatchesFrontendUi/Test/Mftf/Test/StorefrontDisplayCorrectMessageWhenChildProductSelectedForConfigProductTest.xml:
```diff
@@ -0,0 +1,304 @@
+-->
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="StorefrontDisplayCorrectMessageWhenChildProductSelectedForConfigProductTest">
```
Instead of
```
<!--Assign product to shared catalog-->
<comment userInput="Assign product to shared catalog" stepKey="commentAssignProductSharedCatalog"/>
```
you can just use
`<!--Assign product to shared catalog-->`

YuliyaLabudova replied:
Fixed

irenelagno commented on app/code/Magento/InventorySwatchesFrontendUi/composer.json:
```diff
@@ -0,0 +1,21 @@
+    "description": "N/A",
+    "require": {
+        "php": "~7.1.3||~7.2.0||~7.3.0",
+        "magento/framework": "*"
```
should we have a dependency on swatch and inventory modules?

YuliyaLabudova replied:
Yes, I agree. But after adding dependencies static tests failed because we didn't use this functionality. This problem is fixed only if I remove these dependencies

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Fix CR comments.

irenelagno commented on app/code/Magento/InventoryCatalogFrontendUi/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return ResultInterface
+     */
+    public function execute()
```
please add return type

ishakhsuvarov commented on app/code/Magento/InventoryCatalog/Test/Mftf/Data/InventoryData.xml:
```diff
@@ -0,0 +1,22 @@
+
+<entities xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:noNamespaceSchemaLocation="urn:magento:mftf:DataGenerator/etc/dataProfileSchema.xsd">
+    <entity name="StorefrontCatalogInventoryStockThresholdQtyDefaultConfigData">
```
So is it `CatalogInventory` or `InventoryCatalog`?

ishakhsuvarov commented on app/code/Magento/InventoryCatalogFrontendUi/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return ResultInterface
+     */
+    private function getResultForward()
```
```suggestion
    private function getResultForward(): ResultInterface
```

ishakhsuvarov commented on app/code/Magento/InventoryCatalogFrontendUi/Controller/Product/GetQty.php:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        $websiteCode = $this->storeManager->getWebsite()->getCode();
+        $stockId = $this->stockResolver->execute($salesChannel, $websiteCode)->getStockId();
```
What if sales channel is not a website?

ishakhsuvarov commented on app/code/Magento/InventoryCatalogFrontendUi/Model/GetProductQtyLeft.php:
```diff
@@ -0,0 +1,56 @@
+    {
+        $productSalableQty = $this->getProductSalableQty->execute($productSku, $stockId);
+        if ($this->qtyLeftChecker->useQtyForViewing($productSalableQty)) {
+            return  $productSalableQty;
```
```suggestion
            return $productSalableQty;
```

ishakhsuvarov commented on app/code/Magento/InventoryCatalogFrontendUi/Model/GetProductQtyLeft.php:
```diff
@@ -0,0 +1,56 @@
+     * @param int $stockId
+     * @return float|null
+     */
+    public function execute(string $productSku, int $stockId):? float
```
```suggestion
    public function execute(string $productSku, int $stockId): ?float
```

ishakhsuvarov commented on app/code/Magento/InventoryCatalogFrontendUi/Model/QtyLeftChecker.php:
```diff
@@ -0,0 +1,45 @@
+        return ($this->stockItemConfig->getBackorders() === StockItemConfigurationInterface::BACKORDERS_NO
+            || $this->stockItemConfig->getBackorders() !== StockItemConfigurationInterface::BACKORDERS_NO
+            && $this->stockItemConfig->getMinQty() < 0)
+            && $productSalableQty <= $this->stockItemConfig->getStockThresholdQty()
```
Doing precise operations with [floats](https://www.php.net/manual/en/language.types.float.php) may yield [inaccurate results](https://floating-point-gui.de/).

It might be a good idea to take advantage of `[bccomp](https://www.php.net/manual/en/function.bccomp.php)`.

ishakhsuvarov commented on app/code/Magento/InventoryCatalogFrontendUi/Model/QtyLeftChecker.php:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /**
+     * Use qty left for viewing.
```
This comment does not provide too much information

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/README.md:
```diff
@@ -0,0 +1,16 @@
+# InventoryConfigurableProductFrontend module
+
+The `InventoryConfigurableProductFrontend` adds info about qty left quantity for configurable variations.
```
> qty left quantity

 ?

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/i18n/en_US.csv:
```diff
@@ -0,0 +1,2 @@
+"Qty","Qty"
```
Should we use full word `Quantity` for translatable strings?

YuliyaLabudova replied:
I don't use the full word here. I think it is unnecessary.

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/view/frontend/web/js/configurable-variation-qty.js:
```diff
@@ -0,0 +1,45 @@
+    return function (productId) {
+        var selectorInfoStockSkuQty = '.availability.only',
+            selectorInfoStockSkuQtyValue = '.availability.only > strong',
+            salesChannel = 'website',
```
Would be great to have this as a component configuration.

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/view/frontend/web/js/configurable-variation-qty.js:
```diff
@@ -0,0 +1,45 @@
+
+        if (!_.isUndefined(productId) && productId !== null) {
+            $.ajax({
+                url: urlBuilder.build('catalog/product/getQty/'),
```
Why `catalog`? 

Ideally catalog entities should not be aware of the concept of Quantity, as this is a separate domain of inventory.

ishakhsuvarov commented on app/code/Magento/InventoryConfigurableProductFrontend/view/frontend/web/js/configurable-variation-qty.js:
```diff
@@ -0,0 +1,45 @@
+                productQtyInfoBlock.hide();
+            });
+        } else {
+            productQtyInfoBlock.hide();
```
Why isn't it hidden in the beginning?

YuliyaLabudova replied:
It is hidden in the beginning. I do it using css: https://github.com/magento/inventory/blob/0fe82ce945f6794c48912042385f0247e21ad705/app/code/Magento/InventoryConfigurableProductFrontend/view/frontend/web/css/source/_module.less

ishakhsuvarov commented on app/code/Magento/InventorySalesFrontendUi/composer.json:
```diff
@@ -5,6 +5,7 @@
         "php": "~7.1.3||~7.2.0||~7.3.0",
         "magento/framework": "*",
         "magento/module-catalog-inventory": "*",
+        "magento/module-inventory-catalog-frontend-ui": "*",
```
This might not be the best showcase of modularity to have Frontend modules depend on other frontend modules

YuliyaLabudova replied:
We can stay code from the file app/code/Magento/InventorySalesFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php as it was before but in this case, we have code duplicating. Also, I think that it is not the best way to move code from QtyLeftChecker to InventoryCatalog module because it is related only to frontend part.
Could you please suggest anything?

ishakhsuvarov commented on app/code/Magento/InventorySwatchesFrontendUi/README.md:
```diff
@@ -0,0 +1,13 @@
+# InventorySwatchesFrontendUi module
+
+The `InventorySwatchesFrontendUi` adda info about qty left to ui pages.
```
> `adda info`

?

Please keep text descriptive. It's a good place to mention that this module is a part of the feature being developed.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Fix for channel type & code: get it from request params.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Change implementation with id to using sku.

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Add type hinting.

YuliyaLabudova committed:
Merge branch '2.3-develop' of https://github.com/magento/inventory into MC-17694

YuliyaLabudova committed:
MC-17694: Inventory Threshold doesn't display for configurable products simple product

- Fix composer files for composer static tests;
- Rename function that call from another class.

VeronikaYarats committed:
Merge remote-tracking branch 'remotes/msi/2.3-develop' into MC-17694

VeronikaYarats committed:
Merge remote-tracking branch 'remotes/msi/2.3-develop' into MC-17694

YuliyaLabudova committed:
Merge branch '2.3-develop' of https://github.com/magento/inventory into MC-17694

VeronikaYarats committed:
Merge remote-tracking branch 'remotes/msi/2.3-develop' into MC-17694

VeronikaYarats committed:
Merge branch 'MC-17694' of github.com:magento-epam/msi into MC-17694

ishakhsuvarov commented:
@YuliyaLabudova Please merge missing parts into mainline to make sure builds are green on this PR before we can proceed.

YuliyaLabudova committed:
Merge branch '2.3-develop' of https://github.com/magento/inventory into MC-17694

YuliyaLabudova commented:
> @YuliyaLabudova Please merge missing parts into mainline to make sure builds are green on this PR before we can proceed.

I merged mainline. Functional tests failed because we use our changed from ce repository. 
But here this test is green:  http://10.234.156.158:8080/job/Functional-Tests-CE/911/
http://10.234.156.158:8080/job/Functional-Tests-EE/872/
http://10.234.195.187:8080/job/Functional-Tests-B2B/843/

ishakhsuvarov commented:
@YuliyaLabudova Please make sure these changes are not in your repository but in this one.

Thanks.

kieuphan2406 committed:
MC-17694: Merge mainline

smoskaluk commented:
Hi @YuliyaLabudova , looks like this PR doesn't work for configurable products with swatch attributes. May you please recheck?

YuliyaLabudova commented:
> Hi @YuliyaLabudova , looks like this PR doesn't work for configurable products with swatch attributes. May you please recheck?

Hi, unfortunately, now this ticket is processing not by me but by Trigger Team. @kieuphan2406 , could you, please, take a look at it?
From my side, I just can add that we have included the check with swatch attributes to the test scenario for the ticket.

YuliyaLabudova commented:
> Hi @YuliyaLabudova , looks like this PR doesn't work for configurable products with swatch attributes. May you please recheck?

Hi, do you check with branch MC-17694 from the repo https://github.com/magento-epam/magento2ce/ ?
This branch also contains code related to the swatch.

smoskaluk commented:
> > Hi @YuliyaLabudova , looks like this PR doesn't work for configurable products with swatch attributes. May you please recheck?
> 
> Hi, do you check with branch MC-17694 from the repo https://github.com/magento-epam/magento2ce/ ?
> This branch also contains code related to the swatch.

Hi, I don't have access to that repo, I've tested MC-17694 which is here. Any chance for them to be synchronized?

kieuphan2406 committed:
Merge remote-tracking branch 'origin/2.3-develop' into MC-17694

slavvka commented:
@magento run all tests

zakdma committed:
Merge remote-tracking branch 'origin/2.3-develop' into MC-17694

zakdma committed:
Merge remote-tracking branch 'origin/2.3-develop' into MC-17694

sidolov commented:
Hi @YuliyaLabudova  , due to changes in the repository structure (Magento core code is extracted from the MSI code base) we have changed the base branch to `1.1-develop-legacy`. We will keep this branch for a while to make a smooth transition, in the meanwhile, please, apply changes from the current PR to the new code base structure in `1.1-develop` branch. 

[Installation Guide](https://github.com/magento/inventory/wiki/Metapackage-Installation-Guide) may be helpful for local project development. 

Let me know if you need help or have questions!
Thank you for your collaboration!

sidolov commented:
@YuliyaLabudova  I'm closing this pull request since it's targeting legacy branch and cannot be delivered without additional changes because the repository structure has changed. Please, feel free to open the new pull request with the same changes if you wish to continue! Thank you for your collaboration!
